velox/functions/prestosql/GeometryUtils.h

```diff
@@ -0,0 +1,685 @@
+        break;
+      case geos::geom::GEOS_MULTILINESTRING:
+        writePolyline(geometry, sliceOutput, true);
+        break;
+      case geos::geom::GEOS_POLYGON:
+        writePolygon(geometry, sliceOutput, false);
+        break;
+      case geos::geom::GEOS_MULTIPOLYGON:
+        writePolygon(geometry, sliceOutput, true);
+        break;
+      case geos::geom::GEOS_GEOMETRYCOLLECTION:
+        writeGeometryCollection(geometry, sliceOutput);
+        break;
+      default:
+        VELOX_USER_FAIL(
+            "Unrecognized geometry type: {}", static_cast<int>(geometryType));
+        break;
+    }
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> deserialize(
+      SliceInput& sliceInput) {
+    auto geometryType =
+        static_cast<GeometrySerializationType>(sliceInput.readByte());
+    switch (geometryType) {
+      case GeometrySerializationType::POINT:
+        return readPoint(sliceInput);
+      case GeometrySerializationType::MULTI_POINT:
+        return readMultiPoint(sliceInput);
+      case GeometrySerializationType::LINE_STRING:
+        return readPolyline(sliceInput, false);
+      case GeometrySerializationType::MULTI_LINE_STRING:
+        return readPolyline(sliceInput, true);
+      case GeometrySerializationType::POLYGON:
+        return readPolygon(sliceInput, false);
+      case GeometrySerializationType::MULTI_POLYGON:
+        return readPolygon(sliceInput, true);
+      case GeometrySerializationType::ENVELOPE:
+        return readEnvelope(sliceInput);
+      case GeometrySerializationType::GEOMETRY_COLLECTION:
+        return readGeometryCollection(sliceInput);
+      default:
+        VELOX_USER_FAIL(
+            "Unrecognized geometry type: {}",
+            static_cast<uint8_t>(geometryType));
+        break;
+    }
+    return nullptr;
+  }
+
+  static bool isEsriNaN(double d) {
+    auto translateFromAVNaN = [](double value) -> double { return value; };
+    return std::isnan(d) || std::isnan(translateFromAVNaN(d));
+  }
+
+  static bool isClockwise(
+      const std::unique_ptr<geos::geom::CoordinateSequence>& coordinates,
+      int start,
+      int end) {
+    double sum = 0.0;
+    for (int i = start; i < end - 1; i++) {
+      const auto& p1 = coordinates->getAt(i);
+      const auto& p2 = coordinates->getAt(i + 1);
+      sum += (p2.x - p1.x) * (p2.y + p1.y);
+    }
+    return sum > 0.0;
+  }
+
+  static void reverse(
+      const std::unique_ptr<geos::geom::CoordinateSequence>& coordinates,
+      int start,
+      int end) {
+    for (int i = 0; i < (end - start) / 2; ++i) {
+      auto temp = coordinates->getAt(start + i);
+      coordinates->setAt(coordinates->getAt(end - 1 - i), start + i);
+      coordinates->setAt(temp, end - 1 - i);
+    }
+  }
+
+  template <typename T>
+  static void writeEnvelope(
+      const geos::geom::Geometry* geometry,
+      SliceOutput<T>& output) {
+    if (geometry->isEmpty()) {
+      output.writeDouble(std::nan(""));
```

Comment:
just curious, do we need to set an similar magic NAN values here like those that `isEsriNaN` identifies?